### PR TITLE
feat(cart): persist customer carts with stock-aware quantities

### LINKED ISSUE #71: feat(cart): persist customer carts and expose the cart API
## Outcome
Persist one signed-in customer's cart across sessions and expose a documented, ownership-safe FastAPI contract.

## Scope and agreed behavior
- Signed-in carts only. Guest carts and guest-to-account merging are outside this increment.
- One active cart per customer; one line per product. Read, set an absolute quantity, and remove a line. Setting a quantity is idempotent; removal is explicit.
- Quantities are integers from 1 to 99. New lines and quantity increases must not exceed current stock. Adding to a cart does not reserve stock.
- Reads use current backend product prices and calculate line totals/subtotal in exact GBP decimals. Never accept client prices, totals or customer ownership fields.
- If stock falls below a saved quantity, retain the line and report insufficient availability. Permit removal and quantity reductions; do not silently clamp quantities. Checkout will revalidate and reserve inventory separately in #29.
- Define and test deletion behavior so deleting a product never breaks cart reads or leaves orphaned lines.

## Acceptance criteria
- [x] Add reviewed Alembic migration and database constraints for customer/product ownership and uniqueness.
- [x] Resolve the customer from the existing active-user dependency; unauthenticated, expired and disabled-user requests cannot use the cart.
- [x] Implement read, quantity-set and removal endpoints with bounded validation, predictable errors and concurrent-update semantics.
- [x] Cart state survives logout/login and application restarts; customers cannot read or change another customer's lines, including guessed identifiers.
- [x] Return authoritative current prices, exact GBP totals and explicit availability information; document non-reservation and price-change behavior.
- [x] Update FastAPI OpenAPI/Swagger, generated frontend types and API onboarding together.
- [x] Add isolated API tests plus PostgreSQL integration tests for ownership, persistence, duplicate/concurrent requests, invalid quantities, stock changes, product deletion and money calculations.
- [x] Pass required CI and current-head external Codex review before merge; verify the migration through the existing deployment pipeline.

## Dependencies and delivery
Parent: #28. Depends on completed #27. This is the first cart implementation ticket. Use a focused API/persistence PR with `Closes` this ticket and `Refs #28`; do not close the parent until the UI and full journey are delivered.

## Delivery evidence
Merged PR #79 at main `c0e76fd11b12687a93c8d67147ccdc622560e42f`. Required CI and clean Codex review passed on final head `70aa5c1`. Migration and production deployment succeeded: https://github.com/youneshenniwrites/fastapi-next-ecommerce/actions/runs/34634233222. Live OpenAPI exposes cart GET/PUT/DELETE. Repository guides and Wiki updated. Cart storefront remains #72; parent #28 stays open.

### PR DESCRIPTION
**Pull request by youneshenniwrites**
## Summary
Persist each signed-in customer's cart across sessions. Customers can read their cart, set an absolute product quantity, and remove a line; FastAPI calculates current GBP totals and reports stock shortages without silently changing saved quantities.

## Issue
Closes #71. Refs #28. Storefront controls remain the dependent #72.

## Acceptance criteria
- [x] Composite customer/product key, quantity constraint and cascading deletion in migration 0003.
- [x] Active-user ownership, strict quantity validation, idempotent writes and stock rules.
- [x] Current prices, exact GBP totals and explicit availability; no stock reservation.
- [x] OpenAPI, generated frontend types and cart onboarding updated.
- [x] API, persistence, isolation, constraints and PostgreSQL concurrency tests added.
- [ ] PostgreSQL/container CI and current-head external review pass.
- [ ] Deployment applies the migration and passes verification.

## Testing
- Backend lint/format and full suite: 92 passed, 2 PostgreSQL-only skips; coverage 92.5%.
- Additional maximum-money and customer-cascade checks: final cart suite 14 passed, 1 PostgreSQL-only skip.
- Migration upgrade/downgrade/metadata checks passed on disposable SQLite.
- Frontend API generation and typecheck passed.
- Prior-head PostgreSQL and container CI passed; final-head CI is rerunning after documentation follow-up.
- Local Docker daemon unavailable; PostgreSQL row locking and container rollout require CI. No production test data was created.

## Review
Self-review covered ownership, lock ordering, deletion, money precision and migration consistency. CodeRabbit requested explicit migrated-constraint assertions; added and verified in 11a4bff, with the thread resolved. Codex subsequently found stale wording in docs/api.md; corrected the canonical walkthrough and summary in 6f289f9 and resolved the thread. Fresh final-head review is pending.

## Deployment / Compatibility
Additive migration 0003 creates cart_lines. Existing users/products are unchanged. Product deletion intentionally removes its cart lines. Production remains PostgreSQL; SQLite only provides sequential test coverage. Checkout and cart UI are outside this PR.

### CODE REVIEW

**Comment by youneshenniwrites** [comment]
@codex review

**Comment by youneshenniwrites** [comment]
@codex review

File: backend/app/tests/test_migrations.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def alembic(*args):
     alembic("upgrade", "head")
     alembic("check")
     engine = create_engine(database_url)
-    assert {"users", "products"} <= set(inspect(engine).get_table_names())
+    assert {"users", "products", "cart_lines"} <= set(inspect(engine).get_table_names())
```
[the remark this answers is not in the document]
Comment (youneshenniwrites) [reply backend/app/tests/test_migrations.py:33]: Addressed in 11a4bff. The migration test now inspects the composite primary key, both foreign keys and CASCADE actions, and the quantity check name and expression. Both migration tests pass locally, with Ruff lint/format passing; PostgreSQL CI will verify the same assertions on PostgreSQL.

**Comment by youneshenniwrites** [comment]
@codex review

File: backend/app/main.py [hunk 2]
```diff
@@ -85,6 +85,7 @@ async def health_check() -> dict[str, str]:
 # from app.api.v1 import products, users
 app.include_router(products.router, prefix="/api/v1/products", tags=["Products"])
 app.include_router(auth.router, prefix="/api/v1/auth", tags=["auth"])
+app.include_router(cart.router, prefix="/api/v1/cart", tags=["Cart"])
```
[the remark this answers is not in the document]
Comment (youneshenniwrites) [reply backend/app/main.py:88]: Fixed in 6f289f9. docs/api.md now includes the authenticated PUT/GET/DELETE cart walkthrough, 409 stock errors, quantity bounds, current-price/non-reservation behavior and a link to the detailed cart contract. Removed the contradictory no-cart statement. Verified paths and response behavior against generated OpenAPI and the passing cart tests; git diff --check passes.

**Comment by youneshenniwrites** [comment]
@codex review

**Comment by youneshenniwrites** [comment]
@codex review
